scanpipe/models.py

```diff
@@ -1857,6 +1931,193 @@ def update_from_data(self, package_data, override=False):
+        return self.dependency_uid
+
+    def get_absolute_url(self):
+        return reverse("dependency_detail", args=[self.project_id, self.pk])
+
+    @cached_property
+    def for_package_uid(self):
+        if self.for_package:
+            return self.for_package.package_uid
+        else:
+            return ""
+
+    @cached_property
+    def datafile_path(self):
+        if self.datafile_resource:
+            return self.datafile_resource.path
+        else:
+            return ""
+
+    @classmethod
+    def create_from_data(
+        cls,
+        project,
+        dependency_data,
+        for_package=None,
+        datafile_resource=None,
+        strip_datafile_path_root=False,
+    ):
+        """
+        Creates and returns a DiscoveredDependency for a `project` from the
+        `dependency_data`.
+
+        If `strip_datafile_path_root` is True, then `create_from_data()` will
+        strip the root path segment from the `datafile_path` of
+        `dependency_data` before looking up the corresponding CodebaseResource
+        for `datafile_path`. This is used in the case where Dependency data is
+        imported from a scancode-toolkit scan, where the root path segments are
+        not stripped for `datafile_path`s.
+        """
+        required_fields = ["purl", "dependency_uid"]
+        missing_values = [
+            field_name
+            for field_name in required_fields
+            if not dependency_data.get(field_name)
+        ]
+
+        if missing_values:
+            message = (
+                f"No values for the following required fields: "
+                f"{', '.join(missing_values)}"
+            )
+
+            project.add_error(error=message, model=cls, details=dependency_data)
+            return
+
+        if not for_package:
+            for_package_uid = dependency_data.get("for_package_uid")
+            if for_package_uid:
+                for_package = project.discoveredpackages.get(
+                    package_uid=for_package_uid
+                )
+
+        if not datafile_resource:
+            datafile_path = dependency_data.get("datafile_path")
+            if datafile_path:
+                if strip_datafile_path_root:
+                    segments = datafile_path.split("/")
+                    datafile_path = "/".join(segments[1:])
+                datafile_resource = project.codebaseresources.get(path=datafile_path)
+
+        cleaned_dependency_data = {
+            field_name: value
+            for field_name, value in dependency_data.items()
+            if field_name in DiscoveredDependency.model_fields() and value
+        }
+        discovered_dependency = cls(
+            project=project,
+            for_package=for_package,
+            datafile_resource=datafile_resource,
+            **cleaned_dependency_data,
+        )
+        discovered_dependency.save()
+
+        return discovered_dependency
+
+    def update_from_data(self, dependency_data):
```

Comment:
Seems like duplicated code, time for refactoring into a Mixin class. The base code seems generic enough.